lib/utils/reactivation_authenticator.dart

```diff
@@ -0,0 +1,92 @@
+import 'dart:async';
+
+import 'package:chopper/chopper.dart';
+import 'package:coffeecard/cubits/authentication/authentication_cubit.dart';
+import 'package:coffeecard/data/repositories/v1/account_repository.dart';
+import 'package:coffeecard/data/storage/secure_storage.dart';
+import 'package:coffeecard/service_locator.dart';
+
+class ReactivationAuthenticator extends Authenticator {
```

Comment:
Open question: Have you considered the thread-safety of this class? 

I know @TTA777 has introduced parallelism of REST calls which _might_ lead to a scenario where multiple threads end in the authenticator class. This might cause a scenario where read/writes are not visible to the two threads (I think 🤔 )

Reply:
So I just tried to find out if secure_storage is intended to be thread safe, and it would appear so. According to this [issue](https://github.com/mogol/flutter_secure_storage/issues/171) only versions before 3.3.5 should have issues, and we are on 5.0.2. With that said, there exists an [open issue](https://github.com/mogol/flutter_secure_storage/issues/354) on secure_storage where it might be unable to read from storage, if the user restored the app from a backup. So it might be prudent to at least wrap our calls to the storage in a try catch as suggested in that issue

Reply:
there can arise problems when the two calls on the receiptsPage both return 401 where two attempts to refresh the token are made. I have made some changes here (https://github.com/AnalogIO/coffeecard_app/tree/fix/frem/checkBearerToken) that attempt to address this issue

Reply:
I assume you are thinking of the static fields you have introduced in the ReactivationAuthenticator?

I understand why you came up with the idea but it's actually only making it worse as static does not guarantee memory visibility across threads (at least in Java). 
Also and maybe more importantly, the Shiftplanning API also uses an Authenticator. Introducing static fields would share the the retry count and debouncer between both API. We don't want that!

That was a little thread safety rant, sorry 😅

I want to bring some good news though. I have tried reading up on the Dart concurrency model. Dart runs **single threaded** by standard in what is called an _isolate_ (memory is isolated and not shared). My understanding is that unless you explicitly create several isolate to make multi threaded code, the code will always be single threaded.

My first thought is therefore that we don't have a thread issue here and no risk of dirty read/writes.

Do you agree on my understanding? @fremartini @TTA777

[Dart Concurrency documentation](https://dart.dev/guides/language/concurrency)

Reply:
I would like to point out that dart isolates do not work with platform channels in case that information was missed, i.e. you cannot invoke native methods through isolates. Multi-threading will always have to be handled in the native code of the plugin in that scenario.

Reply:
https://github.com/AnalogIO/coffeecard_app/pull/146